server/src/worldgen/schematic.rs

```diff
@@ -0,0 +1,268 @@
+        // TODO: consider a faster algorithm by not checking
+        // for blocks we know are not within the chunk
+
+        let length = self.length as isize;
+        let height = self.height as isize;
+        let width = self.width as isize;
+
+        let center_x = self.center_x as isize;
+        let center_y = self.center_y as isize;
+        let center_z = self.center_z as isize;
+
+        let min_x = -center_x;
+        let min_y = -center_y;
+        let min_z = -center_z;
+
+        let max_x = length - center_x;
+        let max_y = height - center_y;
+        let max_z = width - center_z;
+
+        // Go through each block in the schematic and find
+        // its coordinates local to the chunk.
+        for x in min_x..max_x {
+            for y in min_y..max_y {
+                for z in min_z..max_z {
+                    // Compute absolute position of this block.
+                    let ax = abs_x + x;
+                    let ay = abs_y + y;
+                    let az = abs_z + z;
+
+                    let block = self.block_at(x, y, z);
+                    if block == Block::Air {
+                        continue; // Don't set air blocks
+                    }
+
+                    if ay < 0 || ay >= 256 {
+                        continue;
+                    }
+
+                    if let Some((chunk_x, chunk_z)) = coords_within_chunk(ax, az, chunk.position())
+                    {
+                        chunk.set_block_at(chunk_x, ay as usize, chunk_z, block);
+                    }
+                }
+            }
+        }
+    }
+
+    fn index(&self, x: isize, y: isize, z: isize) -> usize {
+        let x = (x + self.center_x as isize) as usize;
+        let y = (y + self.center_y as isize) as usize;
+        let z = (z + self.center_z as isize) as usize;
+
+        self.unoffsetted_index(x, y, z)
+    }
+
+    fn unoffsetted_index(&self, x: usize, y: usize, z: usize) -> usize {
+        y * self.length * self.width + z * self.length + x
+    }
+}
+
+fn coords_within_chunk(ax: isize, az: isize, chunk: ChunkPosition) -> Option<(usize, usize)> {
+    let ax = ax as i32;
+    let az = az as i32;
+
+    if ax / 16 != chunk.x || az / 16 != chunk.z {
+        return None; // Not within chunk
+    }
+
+    let mut offset_x = (ax - chunk.x * 16).abs();
+    let mut offset_z = (az - chunk.z * 16).abs();
+
+    if ax < 0 {
+        offset_x = 15 - offset_x;
+    }
+    if az < 0 {
+        offset_z = 15 - offset_z;
+    }
+
+    Some((offset_x as usize, offset_z as usize))
+}
+
+/// Builder construct for `Schematic`.
+#[derive(Default)]
+pub struct SchematicBuilder {
+    length: usize,
+    height: usize,
+    width: usize,
+    center_x: usize,
+    center_y: usize,
+    center_z: usize,
+}
+
+impl SchematicBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn with_dimensions(mut self, length: usize, height: usize, width: usize) -> Self {
```

Comment:
Wouldn't it be better to collapse with_dimensions into new. I don't see a case where dimensions 0,0,0 would be used.